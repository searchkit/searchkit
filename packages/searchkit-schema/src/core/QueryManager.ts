export type ValueFilter = {
  identifier: string
  value: string
}

export type RangeFilter = {
  identifier: string
  min: number
  max: number
}

export type DateRangeFilter = {
  identifier: string
  dateMin: string
  dateMax: string
}

export type GeoPoint = {
  lat: number
  lon: number
}

export type GeoBoundingBoxFilter = {
  identifier: string
  geoBoundingBox: {
    topLeft: GeoPoint
    bottomRight: GeoPoint
    topRight: GeoPoint
    bottomLeft: GeoPoint
  }
}

export type MixedFilter = ValueFilter | RangeFilter | DateRangeFilter | GeoBoundingBoxFilter

export default class QueryManager {
  constructor(private filters: Array<MixedFilter>, private query: string) {}

  hasFilters(): boolean {
    return this.filters && this.filters.length > 0
  }

  hasQuery(): boolean {
    return !!(this.query && this.query.length > 0)
  }

  getQuery(): string {
    return this.query
  }

  getFilters(): Array<MixedFilter> {
    return this.hasFilters() ? this.filters : []
  }

  getFiltersById(id: string): Array<MixedFilter> {
    if (!this.hasFilters()) return null
    const idFilters = this.filters.filter((filter) => filter.identifier === id)
    return idFilters.length > 0 ? idFilters : null
  }
}
