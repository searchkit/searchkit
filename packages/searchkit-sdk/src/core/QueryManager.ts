import { SortingOption } from '..'

export type ValueFilter = {
  identifier: string
  value: string
}

export type RangeFilter = {
  identifier: string
  min?: number
  max?: number
}

export type DateRangeFilter = {
  identifier: string
  dateMin?: string
  dateMax?: string
}

export type GeoPoint = {
  lat: number
  lon: number
}

export type GeoBoundingBoxFilter = {
  identifier: string
  geoBoundingBox: {
    topLeft?: GeoPoint
    bottomRight?: GeoPoint
    topRight?: GeoPoint
    bottomLeft?: GeoPoint
  }
}

export type HierarchicalValueFilter = {
  identifier: string
  value: string
  level: number
}

export type QueryOptions = {
  fields: Array<string>
}

export type MixedFilter =
  | ValueFilter
  | RangeFilter
  | DateRangeFilter
  | GeoBoundingBoxFilter
  | HierarchicalValueFilter

export default class QueryManager {
  private sortBy: SortingOption
  private filters: Array<MixedFilter>
  private query: string
  private queryOptions: QueryOptions

  constructor() {
    this.filters = []
    this.query = null
  }

  setQuery(query: string): void {
    this.query = query
  }

  setQueryOptions(options: QueryOptions): void {
    this.queryOptions = options
  }

  setFilters(filters: Array<MixedFilter>): void {
    this.filters = filters
  }

  hasFilters(): boolean {
    return this.filters && this.filters.length > 0
  }

  hasQuery(): boolean {
    return !!(this.query && this.query.length > 0)
  }

  getQuery(): string {
    return this.query
  }

  getQueryOptions(): QueryOptions {
    return this.queryOptions
  }

  getFilters(): Array<MixedFilter> {
    return this.hasFilters() ? this.filters : []
  }

  getFiltersById(id: string): Array<MixedFilter> {
    if (!this.hasFilters()) return null
    const idFilters = this.filters.filter((filter) => filter.identifier === id)
    return idFilters.length > 0 ? idFilters : null
  }

  setSortBy(sort: SortingOption): void {
    this.sortBy = sort
  }

  getSortBy(): SortingOption {
    return this.sortBy
  }
}
