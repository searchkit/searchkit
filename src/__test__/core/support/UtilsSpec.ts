import {Utils} from "../../../"


describe("Utils", ()=> {


  it("guid()", ()=> {
    expect(Utils.guid().length).toEqual(36)
  })

  it("collapse()", ()=> {
    let times2 = n => n*2
    expect(
      Utils.collapse([times2, times2, times2], 1)
    ).toBe(8)
  })

  it("instanceOf", ()=> {
    let isRegex = Utils.instanceOf(RegExp)
    expect(isRegex(/s/)).toBe(true)
    expect(isRegex("s")).toBe(false)
  })

  it("interpolate", ()=> {
    expect(Utils.interpolate("hello {message}, Hello {message}{ending}{missing}", {
      message:"World",
      ending:"!"
    })).toEqual("hello World, Hello World!{missing}")
  })

  it("translate() - don't call interpolate", ()=> {
    let key = "No Results found!"
    spyOn(Utils, "interpolate")
    expect(Utils.translate(key)).toBe(key)
    expect(Utils.interpolate).not.toHaveBeenCalled()
  })

  it("translate() - with interpolations", ()=> {
    let key = "{count} Results found!"
    spyOn(Utils, "interpolate").and.callThrough()
    expect(Utils.translate(key, {count:5})).toBe(
      "5 Results found!"
    )
    expect(Utils.interpolate).toHaveBeenCalledWith(
      key, {count:5}
    )
  })

  it("autobind()", ()=> {
    class Foo {
      message="hello"

      @Utils.autobind
      getMessage(something){
        return this.message + " " + something
      }
    }

    let fn = new Foo().getMessage
    expect(fn("world")).toBe("hello world")

  })
})
