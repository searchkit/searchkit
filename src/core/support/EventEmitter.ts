import {without} from "lodash"
import {each} from "lodash"

export class EventEmitter {
  listeners = []

  addListener(fn){
    this.listeners.push(fn)
    return ()=>{
      this.listeners = without(this.listeners, fn)
    }
  }

  trigger(...args){
    each(this.listeners, (fn)=> {
      fn.apply(null, args)
    })
  }
}
