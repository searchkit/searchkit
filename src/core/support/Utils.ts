import {reduce} from "lodash"
import {map} from "lodash"
import {reject} from "lodash"
import {isUndefined} from "lodash"

export class Utils {
  static guidCounter = 0

  static guid(prefix=""){
    let id = ++Utils.guidCounter
    return prefix.toString() + id
  }

  static collapse(collection, seed){
    const reducer = (current, fn)=> fn(current)
    return reduce(collection, reducer, seed)
  }

  static instanceOf(klass){
    return (val)=> val instanceof klass
  }

  static interpolate(str, interpolations){
    return str.replace(
  		/{([^{}]*)}/g,
  		(a, b) => {
  			var r = interpolations[b];
  			return typeof r === 'string' || typeof r === 'number' ? r : a;
  		}
    )
  }

  static translate(key, interpolations?){
    if(interpolations){
      return Utils.interpolate(key, interpolations)
    } else {
      return key
    }
  }

  static computeOptionKeys(options, fields, defaultKey){
    return map(options, (option)=> {
      return Utils.generateKeyFromFields(option, fields, defaultKey)
    })
  }

  static generateKeyFromFields(ob, fields, defaultKey){
    if(ob.key){
      return ob
    }
    let fieldValues = reject(map(fields, (field: string)=> ob[field]), isUndefined)
    if(fieldValues.length > 0){
      ob.key = fieldValues.join("_")
    } else {
      ob.key = defaultKey
    }
    return ob
  }
}
