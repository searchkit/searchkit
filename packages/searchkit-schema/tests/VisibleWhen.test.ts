import {
  MultiMatchQuery,
  RefinementSelectFacet,
  SearchkitConfig,
  VisibleWhen,
  FacetSelectedRule
} from '@searchkit/sdk'
import nock from 'nock'
import QueryManager from '../src/core/QueryManager'
import { setupTestServer, callQuery } from './support/helper'
import HitsMock from './__mock-data__/FacetsResolver/results.json'

describe('Facets Resolver', () => {
  const customRule = jest.fn((queryManager: QueryManager, ctx: any) => true)
  describe('should return as expected', () => {
    const runQuery = async (gql) => {
      const response = await callQuery({ gql })
      return response
    }

    const config: SearchkitConfig = {
      host: 'http://localhost:9200',
      index: 'movies',
      hits: {
        fields: ['actors', 'writers']
      },
      query: new MultiMatchQuery({ fields: ['actors', 'writers', 'title^4', 'plot'] }),
      facets: [
        new RefinementSelectFacet({ identifier: 'type', field: 'type.raw', label: 'Type' }),
        VisibleWhen(
          [
            new RefinementSelectFacet({
              identifier: 'writers',
              field: 'writers.raw',
              label: 'Writers',
              display: 'override',
              multipleSelect: true
            }),
            new RefinementSelectFacet({
              identifier: 'actors',
              field: 'actors.raw',
              label: 'Actors'
            }),
            new RefinementSelectFacet({
              identifier: 'genres',
              field: 'genres.raw',
              label: 'Genres'
            })
          ],
          [FacetSelectedRule('type', 'Movie'), customRule]
        )
      ]
    }

    it('should return correct Results', async () => {
      setupTestServer({
        config,
        addToQueryType: true,
        typeName: 'ResultSet',
        hitTypeName: 'ResultHit'
      })

      const gql = `
        {
          results(query: "") {
            summary {
              appliedFilters {
                identifier
              }
              disabledFilters {
                identifier
              }
            }
            hits(page: {size: 10, from: 0 }) {
              items {
                id
              }
            }
            facets {
              identifier
              type
              label
              display
              entries {
                count
                label
              }
            }
          }
        }
      `

      const scope = nock('http://localhost:9200')
        .post('/movies/_search')
        .reply((uri, body) => {
          expect(body).toMatchInlineSnapshot(`
            Object {
              "_source": Object {
                "includes": Array [
                  "actors",
                  "writers",
                ],
              },
              "aggs": Object {
                "facet_bucket_all": Object {
                  "aggs": Object {
                    "type": Object {
                      "terms": Object {
                        "field": "type.raw",
                        "size": 5,
                      },
                    },
                  },
                  "filter": Object {
                    "bool": Object {
                      "must": Array [],
                    },
                  },
                },
              },
              "from": 0,
              "size": 10,
              "sort": Array [
                Object {
                  "_score": "desc",
                },
              ],
            }
          `)
          return [200, HitsMock]
        })

      const response = await runQuery(gql)
      expect(response.body.data).toMatchSnapshot()
      expect(response.status).toEqual(200)
    })

    it('Reeturn all facets when rule satisfied', async () => {
      setupTestServer({
        config,
        addToQueryType: true,
        typeName: 'ResultSet',
        hitTypeName: 'ResultHit'
      })

      const gql = `
        {
          results(query: "", filters: [{identifier: "type", value:"Movie"}]) {
            summary {
              appliedFilters {
                identifier
              }
              disabledFilters {
                identifier
              }
            }
            hits(page: {size: 10, from: 0 }) {
              items {
                id
              }
            }
            facets {
              identifier
              type
              label
              display
              entries {
                count
                label
              }
            }
          }
        }
      `

      const scope = nock('http://localhost:9200')
        .post('/movies/_search')
        .reply((uri, body) => {
          expect(body).toMatchInlineSnapshot(`
            Object {
              "_source": Object {
                "includes": Array [
                  "actors",
                  "writers",
                ],
              },
              "aggs": Object {
                "facet_bucket_all": Object {
                  "aggs": Object {
                    "actors": Object {
                      "terms": Object {
                        "field": "actors.raw",
                        "size": 5,
                      },
                    },
                    "genres": Object {
                      "terms": Object {
                        "field": "genres.raw",
                        "size": 5,
                      },
                    },
                    "type": Object {
                      "terms": Object {
                        "field": "type.raw",
                        "size": 5,
                      },
                    },
                  },
                  "filter": Object {
                    "bool": Object {
                      "must": Array [
                        Object {
                          "bool": Object {
                            "must": Array [
                              Object {
                                "term": Object {
                                  "type.raw": "Movie",
                                },
                              },
                            ],
                          },
                        },
                      ],
                    },
                  },
                },
                "facet_bucket_writers": Object {
                  "aggs": Object {
                    "writers": Object {
                      "terms": Object {
                        "field": "writers.raw",
                        "size": 5,
                      },
                    },
                  },
                  "filter": Object {
                    "bool": Object {
                      "must": Array [
                        Object {
                          "bool": Object {
                            "must": Array [
                              Object {
                                "term": Object {
                                  "type.raw": "Movie",
                                },
                              },
                            ],
                          },
                        },
                      ],
                    },
                  },
                },
              },
              "from": 0,
              "post_filter": Object {
                "bool": Object {
                  "must": Array [
                    Object {
                      "bool": Object {
                        "must": Array [
                          Object {
                            "term": Object {
                              "type.raw": "Movie",
                            },
                          },
                        ],
                      },
                    },
                  ],
                },
              },
              "size": 10,
              "sort": Array [
                Object {
                  "_score": "desc",
                },
              ],
            }
          `)
          return [200, HitsMock]
        })

      const response = await runQuery(gql)
      expect(response.body.data).toMatchSnapshot()
      expect(response.status).toEqual(200)
    })

    it('should not have the writers filter applied', async () => {
      setupTestServer({
        config,
        addToQueryType: true,
        typeName: 'ResultSet',
        hitTypeName: 'ResultHit'
      })

      const gql = `
        {
          results(query: "", filters: [{ identifier: "writers", value: "example"}]) {
            summary {
              appliedFilters {
                identifier
              }
              disabledFilters {
                identifier
              }
            }
            hits(page: {size: 10, from: 0 }) {
              items {
                id
              }
            }
            facets {
              identifier
              entries {
                label
              }
            }
          }
        }
      `

      const scope = nock('http://localhost:9200')
        .post('/movies/_search')
        .reply((uri, body) => {
          expect(body).toMatchInlineSnapshot(`
            Object {
              "_source": Object {
                "includes": Array [
                  "actors",
                  "writers",
                ],
              },
              "aggs": Object {
                "facet_bucket_all": Object {
                  "aggs": Object {
                    "type": Object {
                      "terms": Object {
                        "field": "type.raw",
                        "size": 5,
                      },
                    },
                  },
                  "filter": Object {
                    "bool": Object {
                      "must": Array [],
                    },
                  },
                },
              },
              "from": 0,
              "size": 10,
              "sort": Array [
                Object {
                  "_score": "desc",
                },
              ],
            }
          `)
          return [200, HitsMock]
        })

      const response = await runQuery(gql)
      expect(response.body.data).toMatchSnapshot()
      expect(response.status).toEqual(200)
    })

    it('custom rule configuration', async () => {
      setupTestServer({
        config,
        addToQueryType: true,
        typeName: 'ResultSet',
        hitTypeName: 'ResultHit'
      })

      const gql = `
        {
          results(query: "", filters: [{identifier: "type", value:"Movie"}]) {
            facets {
              identifier
            }
          }
        }
      `

      const scope = nock('http://localhost:9200')
        .post('/movies/_search')
        .reply((uri, body) => [200, HitsMock])

      const response = await runQuery(gql)
      expect(response.body.data).toMatchSnapshot()
      expect(response.status).toEqual(200)
      const customRuleArgs = customRule.mock.calls[0]
      expect(customRuleArgs[0].getQuery()).toBe(null)
    })
  })
})
