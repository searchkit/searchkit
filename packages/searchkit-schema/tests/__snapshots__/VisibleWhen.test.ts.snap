// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Facets Resolver should return as expected Reeturn all facets when rule satisfied 2`] = `
Object {
  "results": Object {
    "facets": Array [
      Object {
        "display": "ListFacet",
        "entries": Array [
          Object {
            "count": 3120,
            "label": "Movie",
          },
          Object {
            "count": 521,
            "label": "Episode",
          },
          Object {
            "count": 473,
            "label": "Series",
          },
          Object {
            "count": 48,
            "label": "Game",
          },
        ],
        "identifier": "type",
        "label": "Type",
        "type": "RefinementSelectFacet",
      },
      Object {
        "display": "override",
        "entries": Array [
          Object {
            "count": 83,
            "label": "J.J. Abrams",
          },
          Object {
            "count": 74,
            "label": "Jeffrey Lieber",
          },
          Object {
            "count": 73,
            "label": "Damon Lindelof",
          },
          Object {
            "count": 53,
            "label": "James Manos Jr.",
          },
          Object {
            "count": 53,
            "label": "Jeff Lindsay",
          },
        ],
        "identifier": "writers",
        "label": "Writers",
        "type": "RefinementSelectFacet",
      },
      Object {
        "display": "ListFacet",
        "entries": Array [
          Object {
            "count": 73,
            "label": "Naveen Andrews",
          },
          Object {
            "count": 56,
            "label": "Jennifer Carpenter",
          },
          Object {
            "count": 56,
            "label": "Michael C. Hall",
          },
          Object {
            "count": 53,
            "label": "Emilie de Ravin",
          },
          Object {
            "count": 42,
            "label": "Jared Padalecki",
          },
        ],
        "identifier": "actors",
        "label": "Actors",
        "type": "RefinementSelectFacet",
      },
      Object {
        "display": "ListFacet",
        "entries": Array [
          Object {
            "count": 2529,
            "label": "Drama",
          },
          Object {
            "count": 1274,
            "label": "Comedy",
          },
          Object {
            "count": 786,
            "label": "Crime",
          },
          Object {
            "count": 725,
            "label": "Action",
          },
          Object {
            "count": 666,
            "label": "Adventure",
          },
        ],
        "identifier": "genres",
        "label": "Genres",
        "type": "RefinementSelectFacet",
      },
    ],
    "hits": Object {
      "items": Array [
        Object {
          "id": "tt0468569",
        },
        Object {
          "id": "tt0110912",
        },
        Object {
          "id": "tt0167260",
        },
        Object {
          "id": "tt0114369",
        },
        Object {
          "id": "tt0073486",
        },
        Object {
          "id": "tt0172495",
        },
        Object {
          "id": "tt1520211",
        },
        Object {
          "id": "tt0338013",
        },
        Object {
          "id": "tt0078788",
        },
        Object {
          "id": "tt0180093",
        },
      ],
    },
    "summary": Object {
      "appliedFilters": Array [
        Object {
          "identifier": "type",
        },
      ],
      "disabledFilters": Array [],
    },
  },
}
`;

exports[`Facets Resolver should return as expected custom rule configuration 1`] = `
Object {
  "results": Object {
    "facets": Array [
      Object {
        "identifier": "type",
      },
      Object {
        "identifier": "writers",
      },
      Object {
        "identifier": "actors",
      },
      Object {
        "identifier": "genres",
      },
    ],
  },
}
`;

exports[`Facets Resolver should return as expected should not have the writers filter applied 2`] = `
Object {
  "results": Object {
    "facets": Array [
      Object {
        "entries": Array [
          Object {
            "label": "Movie",
          },
          Object {
            "label": "Episode",
          },
          Object {
            "label": "Series",
          },
          Object {
            "label": "Game",
          },
        ],
        "identifier": "type",
      },
    ],
    "hits": Object {
      "items": Array [
        Object {
          "id": "tt0468569",
        },
        Object {
          "id": "tt0110912",
        },
        Object {
          "id": "tt0167260",
        },
        Object {
          "id": "tt0114369",
        },
        Object {
          "id": "tt0073486",
        },
        Object {
          "id": "tt0172495",
        },
        Object {
          "id": "tt1520211",
        },
        Object {
          "id": "tt0338013",
        },
        Object {
          "id": "tt0078788",
        },
        Object {
          "id": "tt0180093",
        },
      ],
    },
    "summary": Object {
      "appliedFilters": Array [],
      "disabledFilters": Array [
        Object {
          "identifier": "writers",
        },
      ],
    },
  },
}
`;

exports[`Facets Resolver should return as expected should return correct Results 2`] = `
Object {
  "results": Object {
    "facets": Array [
      Object {
        "display": "ListFacet",
        "entries": Array [
          Object {
            "count": 3120,
            "label": "Movie",
          },
          Object {
            "count": 521,
            "label": "Episode",
          },
          Object {
            "count": 473,
            "label": "Series",
          },
          Object {
            "count": 48,
            "label": "Game",
          },
        ],
        "identifier": "type",
        "label": "Type",
        "type": "RefinementSelectFacet",
      },
    ],
    "hits": Object {
      "items": Array [
        Object {
          "id": "tt0468569",
        },
        Object {
          "id": "tt0110912",
        },
        Object {
          "id": "tt0167260",
        },
        Object {
          "id": "tt0114369",
        },
        Object {
          "id": "tt0073486",
        },
        Object {
          "id": "tt0172495",
        },
        Object {
          "id": "tt1520211",
        },
        Object {
          "id": "tt0338013",
        },
        Object {
          "id": "tt0078788",
        },
        Object {
          "id": "tt0180093",
        },
      ],
    },
    "summary": Object {
      "appliedFilters": Array [],
      "disabledFilters": Array [],
    },
  },
}
`;
